src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/DefaultRazorSemanticTokensInfoService.cs

```diff
@@ -13,32 +15,33 @@
 using Microsoft.AspNetCore.Razor.LanguageServer.Common;
 using Microsoft.AspNetCore.Razor.LanguageServer.ProjectSystem;
 using Microsoft.AspNetCore.Razor.LanguageServer.Semantic.Models;
-using Microsoft.AspNetCore.Razor.LanguageServer.Semantic.Services;
 using Microsoft.CodeAnalysis;
 using Microsoft.CodeAnalysis.Razor;
 using Microsoft.CodeAnalysis.Razor.ProjectSystem;
 using Microsoft.CodeAnalysis.Razor.Workspaces;
 using Microsoft.Extensions.Logging;
+using OmniSharp.Extensions.LanguageServer.Protocol;
 using OmniSharp.Extensions.LanguageServer.Protocol.Models;
 using OmniSharp.Extensions.LanguageServer.Protocol.Models.Proposals;
 
 namespace Microsoft.AspNetCore.Razor.LanguageServer.Semantic
 {
     internal class DefaultRazorSemanticTokensInfoService : RazorSemanticTokensInfoService
     {
-        // These caches are created to restrict memory growth.
-        // We need to keep track of the last couple of requests for use in previousResultId,
-        // but if we let them grow unbounded it could quickly allocate a lot of memory.
-        // Solution: in-memory caches
-        private readonly MemoryCache<string, (VersionStamp SemanticVersion, IReadOnlyList<int> Data)> _semanticTokensCache = new(); // For Razor docs
-        private readonly MemoryCache<string, IReadOnlyList<int>> _csharpGeneratedSemanticTokensCache = new(); // For C# generated docs
-
         private readonly ClientNotifierServiceBase _languageServer;
+        private readonly RazorDocumentMappingService _documentMappingService;
         private readonly ProjectSnapshotManagerDispatcher _projectSnapshotManagerDispatcher;
         private readonly DocumentResolver _documentResolver;
         private readonly DocumentVersionCache _documentVersionCache;
         private readonly ILogger _logger;
-        private readonly RazorDocumentMappingService _documentMappingService;
+
+        #region Caching constructs
+        // Maps (docURI -> (resultId -> tokens)). We cache per-doc instead of storing all tokens
+        // in one giant cache to improve colorization speeds when working with multiple files.
+        private const int MaxCachesPerDoc = 6;
```

Comment:
It's pretty arbitrary 😄  Roslyn has their cache set to 5 (also arbitrary), so I thought it would be good to pick a similar number for cohesion's sake. The reason it's 6 instead of 5 is because we're using a MemoryCache which halves when full, so I figured an even number would be better.